include/graphblas/reference/blas3.hpp

```diff
@@ -918,6 +937,301 @@ namespace grb {
+						const auto j = A_raw.row_index[ k ];
+
+						// Increment the mask pointer until we find the right column, or a lower column (since the storage withing a row is sorted in a descending order)
+						while( mask_k < mask_raw.col_start[ i + 1 ] && mask_raw.row_index[ mask_k ] > j  ) {
+							_DEBUG_PRINT( "NEquals masked coordinate: ( " + std::to_string( i ) + ";" + std::to_string( mask_raw.row_index[ mask_k ] ) + " )\n" );
+							mask_k++;
+						}
+						if( mask_k >= mask_raw.col_start[ i + 1 ] ) {
+							_DEBUG_PRINT( "No value left for this column\n" );
+							break;
+						}
+
+						if( mask_raw.row_index[ mask_k ] < j ) {
+							_DEBUG_PRINT( "Skip masked coordinate at: ( " + std::to_string( i ) + ";" + std::to_string( mask_raw.row_index[ mask_k ] ) + " )\n" );
+							continue;
+						}
+
+						if( not ignore_mask_values && not MaskHasValue< MaskType >( mask_raw, mask_k ).value ) {
+							_DEBUG_PRINT( "Skip masked value at: ( " + std::to_string( i ) + ";" + std::to_string( mask_raw.row_index[ mask_k ] ) + " )\n" );
+							continue;
+						}
+
+						identity_element_computed |= ( ( descr & descriptors::add_identity ) && i == j );
+
+						// descriptors::add_identity logic
+						const InputType identity_increment = (descr & descriptors::add_identity) && i == j ? static_cast<InputType>(1) : static_cast<InputType>(0);
+						_DEBUG_PRINT( "identity_increment = " + std::to_string( identity_increment ) + "\n" );
+
+						// Get A value
+						const InputType a_val = A_raw.getValue( k, identity );
+						_DEBUG_PRINT( "A( " + std::to_string( i ) + ";" + std::to_string( j ) + " ) = " + std::to_string( a_val ) + "\n" );
+
+						// Compute the fold for this coordinate
+						auto local_x_before = local_x;
+						local_rc = local_rc ? local_rc : grb::apply< descr >( local_x, local_x_before, a_val + identity_increment, op );
+						_DEBUG_PRINT( "Computing: local_x = op(" + std::to_string( a_val + identity_increment ) + ", " + std::to_string( local_x_before ) + ") = " + std::to_string( local_x ) + "\n" );
+					}
+					if( (descr & descriptors::add_identity) && not identity_element_computed && i < n && i < m ) {
+						const auto local_x_before = local_x;
+						local_rc = local_rc ? local_rc : grb::apply< descr >( local_x, local_x_before, static_cast<InputType>(1), op );
+						_DEBUG_PRINT( "Computing identity: local_x = op(" + std::to_string( 1 ) + ", " + std::to_string( local_x_before ) + ") = " + std::to_string( local_x ) + "\n" );
+					}
+				}
+
+#ifdef _H_GRB_REFERENCE_OMP_BLAS3
+	#pragma omp critical
+#endif
+				{ // Reduction with the global result (critical section if OpenMP)
+					auto x_before = x;
+					local_rc = local_rc ? local_rc : grb::apply< descr >( x, x_before, local_x, op );
+					_DEBUG_PRINT( "Computing x: op(" + std::to_string( local_x ) + ", " + std::to_string( x_before ) + ") = " + std::to_string( x ) + "\n" );
+					rc = rc ? rc : local_rc;
+				}
+			}
```

Comment:
similar comments here -- I can review again in more details once you made an initial pass?